benchmark/multidimensional_memset_benchmark_kernels.cpp

```diff
@@ -0,0 +1,24 @@
+
+// Copyright (c) 2023 NVIDIA Corporation (reply-to: brycelelbach@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <flux/op/cartesian_product.hpp>
+#include <flux/source/iota.hpp>
+#include <flux/op/for_each.hpp>
+
+void memset_2d_reference(double* A, std::size_t N, std::size_t M)
+{
+    for (std::size_t j = 0; j != M; ++j)
+        for (std::size_t i = 0; i != N; ++i)
+            A[i + j * N] = 0.0;
+}
+
+void memset_2d_flux_cartesian_product_iota(double* A, std::size_t N, std::size_t M)
+{
+    flux::cartesian_product(flux::iota(0LU, N), flux::iota(0LU, M))
+        .for_each(flux::unpack([&] (auto i, auto j) {
+            A[i + j * N] = 0.0;
+        }));
+}
```

Comment:
This still doesn't work as intend - LLVM fails to vectorize it, leading this code to be 10x slower than it's C-style counterpart.

However, I'm pretty sure we can eventually crack this. The next thing I'm going to try is a separate customization path for `for_each`.

Reply:
This is fixed as of 4132087e2b68de4ad69c0643e1cbaff2b67b5ecd. The flux variant was iterating in the opposite order, leading to strided memory accesses that couldn't be vectorized on some platforms. Now that both the reference and the flux variant are doing the same thing, we get identical assembly.